packages/api/src/car.js

```diff
@@ -303,48 +301,32 @@ async function carStat (carBytes) {
   if (!rawRootBlock) {
     throw new Error('missing root block')
   }
-  if (blocks === 1) {
-    const decoder = decoders.find(d => d.code === rootCid.code)
-    // if we can't decode, we can't check this...
-    if (decoder) {
-      const rootBlock = new Block({ cid: rootCid, bytes: rawRootBlock.bytes, value: decoder.decode(rawRootBlock.bytes) })
-      const numLinks = Array.from(rootBlock.links()).length
-      // if the root block has links, then we should have at least 2 blocks in the CAR
-      if (numLinks > 0) {
-        throw new Error('CAR must contain at least one non-root block')
-      }
+  const decoder = decoders.find(d => d.code === rootCid.code)
+  if (decoder) {
+    const rootBlock = new Block({ cid: rootCid, bytes: rawRootBlock.bytes, value: decoder.decode(rawRootBlock.bytes) })
+    const links = Array.from(rootBlock.links())
+    // if the root block has links, then we should have at least 2 blocks in the CAR
+    if (blocks === 1 && links.length > 0) {
+      throw new Error('CAR must contain at least one non-root block')
+    }
+    // get the size of the full dag for this root, even if we only have a partial CAR.
+    if (rootBlock.cid.code === pb.code) {
+      size = cumulativeSize(rootBlock.bytes, rootBlock.value)
     }
   }
   return { size, blocks }
 }
 
 /**
- * Returns the DAG size of the CAR but only if the graph is complete.
- * @param {Uint8Array} carBytes
+ * The sum of the node size and size of each link
+ * @param {import('@ipld/dag-pb/src/interface').PBNode} pbNode
+ * @returns {number} the size of the DAG in bytes
  */
-async function getDagSize (carBytes) {
-  const reader = await CarReader.fromBytes(carBytes)
-  const [rootCid] = await reader.getRoots()
-
-  const getBlock = async cid => {
-    const rawBlock = await reader.get(cid)
-    if (!rawBlock) throw new Error(`missing block for ${cid}`)
-    const { bytes } = rawBlock
-    const decoder = decoders.find(d => d.code === cid.code)
-    if (!decoder) throw new Error(`missing decoder for ${cid.code}`)
-    return new Block({ cid, bytes, value: decoder.decode(bytes) })
-  }
-
-  const getSize = async cid => {
-    const block = await getBlock(cid)
-    let size = block.bytes.byteLength
-    for (const [, cid] of block.links()) {
-      size += await getSize(cid)
-    }
-    return size
-  }
-
-  return getSize(rootCid)
+function cumulativeSize (pbNodeBytes, pbNode) {
+  // NOTE: Tsize is optional, but all ipfs implementations we know of set it.
+  // It's metadata, that could be missing or deliberately set to an incorrect value.
+  // This logic is the same as used by go/js-ipfs to display the cumulative size of a dag-pb dag.
+  return pbNodeBytes.length + pbNode.Links.reduce((acc, curr) => acc + (curr.Tsize || 0), 0)
```

Comment:
@vasco-santos uh oh. i think this should be:
```diff
-  return pbNodeBytes.length + pbNode.Links.reduce((acc, curr) => acc + (curr.Tsize || 0), 0)
+  return pbNodeBytes.byteLength + pbNode.Links.reduce((acc, curr) => acc + (curr.Tsize || 0), 0)
```

Reply:
Any suggestions to reliably test this?